### PR TITLE
Lua AI changes

### PR DESCRIPTION
**Pull request by mattsc**
This PR implements more Lua AI tasks from issue #3450. Unlike #3506, it includes functionality changes. This is just the beginning, I'll add to it as I work my way through the tasks, but I figured I open this while there is not much yet, rather than overwhelming everybody with a gazillion commits all at once.

### CODE REVIEW

**Comment by CelticMinstrel** [comment]
The proper way to deprecate Lua functions is to use the `deprecate_api` function:

https://github.com/wesnoth/wesnoth/blob/master/data/lua/core.lua#L183

Basically, define your function, then add `your_function = wesnoth.deprecate_api('your_function', 'name of replacement for auto-generated message', level, version, your_function, 'optional extra message')`.

For the case where there's no replacement, it's fine to do it as you did in 01dd850, though I think a replacement for the xyoff function might already exist in the `wesnoth.map` table, so maybe check that first.

**Comment by mattsc** [comment]
@CelticMinstrel  I made the changes to the deprecations a couple days ago, but apparently github does not send messages when you force push.

**Comment by CelticMinstrel** [comment]
I kinda wish we could just write `wesnoth.require "ai/ai_helper"` and it would automatically insert the "lua" directory...

I think `wesnoth.require "ai/lua/ai_helper"` should already work though. If not, that's a bug that needs to be fixed (it's supposed to work).

**Comment by mattsc** [comment]
FTR, I am not planning to add [avoid] tag functionality to any of the other MAIs any more, as it either doesn't make much sense or can already be done (or at least something very similar) with the existing options.  In my opinion.  Let me know if you think it should still be added somewhere else.

**Comment by mattsc** [comment]
> Hmm... isn't the reachmap something that's actually exposed from C++? Not sure if it's possible to rewrite your Lua function in terms of that, though...

It is, that's what wesnoth.find_reach is, unless I am misunderstanding you.  It's not quite the same reachmap though as that also includes hexes occupied by allied units.  Well, and I am turning it into a location set, but that's a minor technical detail.

> It's interesting that you have to go to all that trouble to see if an avoid aspect was explicitly set, but I don't think there's any other way to do it. From the AI framework's perspective, all aspects are always set, usually to a composite aspect containing a default aspect. Still, if you think you might need to do this sort of thing often, perhaps it's worth wrapping the functionality in a separate function...

I probably don't need it in other places.  So far, [avoid] is the only aspect for which I have a use, I think.  But yeah, if that changes we should probably do so.

**Comment by CelticMinstrel** [comment]
You are indeed misunderstanding me. I'm referring to the incredibly-poorly-documented `ai.get_dst_src()`, `ai.get_src_dst()`, `ai.get_enemy_dst_src()`, and `ai.get_enemy_src_dst()`.

**Comment by mattsc** [comment]
> You are indeed misunderstanding me. I'm referring to the incredibly-poorly-documented `ai.get_dst_src()`, `ai.get_src_dst()`, `ai.get_enemy_dst_src()`, and `ai.get_enemy_src_dst()`. (Which are themselves just Lua wrappers around an even worse and equally-poorly-documented Lua API exposed from C++.)

Oh, those.  I actually know those quite well, played with them a lot in the early days.  If even wrote [ai_helper functions such as this](https://github.com/wesnoth/wesnoth/blob/master/data/ai/lua/ai_helper.lua#L1186) in the early days when I was trying to duplicate WFL syntax in Lua.  But they have the same "problem" as `wesnoth.find_reach` for this purpose.

**Comment by CelticMinstrel** [comment]
I wonder if there's any performance benefit of using the C++ reachmaps and then filtering out ally-occupied units in a separate step...

**Comment by mattsc** [comment]
@CelticMinstrel Is b803b08  what you had in mind when you asked for adding named locations to the MAIs in #3450 ?  If so, I'll add it to the other MAIs as well.

**Comment by CelticMinstrel** [comment]
Yes, that's exactly what I meant by supporting named locations. The only other thing is that you might want to consider what happens if the provided location does not exist - I think you'll get a `nil` in that case

**Comment by GregoryLundberg** [comment]
If I do anything (not just for an AI) using a named location which does not exist, I would expect a message to appear on stderr (at least) warning me of the fact. Similarly, if I use a map coordinate which does not exist, I would expect a message. Furthermore, most places where I could reference a map location, I would expect to be able to use a named location instead.

**Comment by mattsc** [comment]
My, my, all those expectations!  ;-)  But yes, I'd already changed my mind on this myself while working on it.  How's this:
```
function ai_helper.get_named_loc_xy(param_core, cfg, required_for)
    -- Get coordinates for either a named location or from x/y coordinates specified
    -- in @cfg. The location can be provided:
    --   - as name: cfg[param_core .. '_loc_name'] (string)
    --   - or as coordinates: cfg[param_core .. '_x'] and cfg[param_core .. '_y'] (integers)
    -- This is the syntax used by many Micro AIs.
    --
    -- Error messages are displayed if the named location does not exist, or if
    -- the coordinates are not on the map. In addition, if @required_for (a string)
    -- is provided, an error message is also displayed if neither a named location
    -- nor both coordinates are provided. If @required_for is not passed and neither
    -- exist, nil is returned.

    local loc
    local loc_name = cfg[param_core .. '_loc_name']
    if loc_name then
        local loc = wesnoth.special_locations[loc_name]
        if loc then
            return loc
        else
            H.wml_error("Named location does not exist: " .. loc_name)
        end
    end

    local x, y = cfg[param_core .. '_x'], cfg[param_core .. '_y']
    if x and y then
        local width, height = wesnoth.get_map_size()
        if (x < 1) or (x > width) or (y < 1) or (y > height) then
            H.wml_error("Location is not on map: " .. param_core .. '_x' .. ',' .. param_core .. '_y = ' .. x .. ',' .. y)
        end

        return { x, y }
    end

    if required_for then
        H.wml_error(required_for .. " requires either " .. param_core .. "_loc_name= or " .. param_core .. "_x/y= keys")
    end
end
```

**Comment by mattsc** [comment]
Since force pushing doesn't send a notification: 8d96107 is what I have now for the return guardian. If that looks okay I'll make the equivalent changes to the other MAIs.

**Comment by CelticMinstrel** [comment]
I think `param_loc_name` is a bit long; I preferred `param_loc` (and in the envoy AI you gave me for BCoA, I used just `param` with no suffix, which I prefer even more).

However, if you _really __must___ make the key a bit more descriptive, I'd go with `param_loc_id` rather than `param_loc_name`, as the key `location_id` is used in other places for named locations, so it strikes me as more consistent. (Obviously you can't use `location_id` here since there could me multiple named locations in the config.)

**Comment by CelticMinstrel** [comment]
Hmm... I could maybe suggest (for consistency with other places) using `location_id` instead of `loc` for the case of an empty `param_core`, at least in the single case. Not sure if that works in the multi case?

**Comment by CelticMinstrel** [comment]
Also, seeing the way you handled the healthy trait, I also want to point out that `u.traits` exists and gives you a list of the trait IDs. I imagine that could be faster than dipping into the config.

**Comment by mattsc** [comment]
> Hmm... I could maybe suggest (for consistency with other places) using `location_id` instead of `loc` for the case of an empty `param_core`

Hmm...  I was going for consistency in the naming across the Micro AI uses, but sure, makes sense.

> that u.traits exists and gives you a list of the trait IDs.

It does?  (Yes, it does, I just tested it.)  That's not documented on the wiki, I specifically looked before doing this, so we should add that.  Yes, that is definitely faster.

Both changes made and squashed in dcf9514 and 5e0df80 .

File: data/ai/micro_ais/cas/ca_fast_attack_utils.lua [hunk 1]
```diff
@@ -184,7 +191,7 @@ function ca_fast_attack_utils.damage_rating_unit(attacker_info, defender_info, a
     -- Input parameters:
     --  @attacker_info, @defender_info: unit_info tables produced by ca_fast_attack_utils.get_unit_info()
     --  @att_stat, @def_stat: attack statistics for the two units
-    --  @is_village: whether the hex from which the attacker attacks is a village
+    --  @healing: the healing given by a village
```
Comment (CelticMinstrel) [inline data/ai/micro_ais/cas/ca_fast_attack_utils.lua]: Should this actually be the healing given by the hex the attacker is on?

Comment (mattsc) [reply data/ai/micro_ais/cas/ca_fast_attack_utils.lua]: Yeah, that's really a better description of what the code does.  I'll correct the comment.

I guess the exception would be if there are hexes on the map that provide healing but that don't count as villages. Does something like that exist (other than hexes next to healers)? I don't actually know how `wesnoth.get_villages()` works internally. But either way, given that this is supposed to be the _fast_ MAI, I would rather not add additional complexity to it.

**Comment by mattsc** [comment]
As far as I am concerned, this is all I wanted to do about the Lua AIs, so I removed the WIP label from the title.  I'll let this sit for a couple days to see if there are more comments, and then merge it.

**Comment by CelticMinstrel** [comment]
I think `wesnoth.get_villages()` only returns terrain spaces that give income, so oases (which heal but do not give income) would be excluded. But I haven't checked so I could be wrong on that.

**Comment by mattsc** [comment]
Yeah, that's correct, I just ran a quick test.  I'll do some speed tests tomorrow to see how I want to deal with that.

**Comment by CelticMinstrel** [comment]
For the record, I have no idea whether it even _should_ be considering oases… I'd say the main AI should, but it probably already does?

**Comment by mattsc** [comment]
It depends. There are rating contributions for healing and others for income aspects of hexes in some Lua AIs.  Currently those are only given for villages, but it should be easy to separate which should be applied when and use the correct metric.

I actually don't know how the default AI does it.  I'll check that when I get to the C++ part of #3450

**Pull request merged**